File: src/schemas/career_profile.py
"""Typed read-only Career Profile contract.

These models define the bounded shape of career data that can be truthfully
derived from current canonical storage. They are used by the API response layer
only; no client writes are accepted in this contract.
"""
from __future__ import annotations

from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class ProvenanceState(str, Enum):
    """Canonical provenance for preview/extracted career items."""

    EXTRACTED_FROM_CV = "extracted_from_cv"


class BaseCareerItem(BaseModel):
    """Common server-owned fields for every piece of career evidence."""

    model_config = {"extra": "ignore"}

    id: Optional[str] = None
    source_document_id: Optional[str] = None
    provenance: ProvenanceState = ProvenanceState.EXTRACTED_FROM_CV
    confidence: Optional[float] = None
    confirmed_at: Optional[str] = None
    updated_at: Optional[str] = None


class ExperienceItem(BaseCareerItem):
    """One work experience entry, derived from CV extraction."""

    role: Optional[str] = Field(None, max_length=200)
    company: Optional[str] = Field(None, max_length=200)
    start_date: Optional[str] = Field(None, max_length=20)
    end_date: Optional[str] = Field(None, max_length=20)
    description: Optional[str] = Field(None, max_length=2000)
    location: Optional[str] = Field(None, max_length=200)


class EducationItem(BaseCareerItem):
    """One education entry, derived from CV extraction."""

    institution: Optional[str] = Field(None, max_length=200)
    degree: Optional[str] = Field(None, max_length=100)
    field: Optional[str] = Field(None, max_length=100)
    start_date: Optional[str] = Field(None, max_length=20)
    end_date: Optional[str] = Field(None, max_length=20)


class SkillItem(BaseModel):
    """One skill as stored in canonical legacy profile.skills."""

    model_config = {"extra": "ignore"}

    name: Optional[str] = Field(None, max_length=100)


class CertificationItem(BaseModel):
    """One certification as stored in canonical legacy profile.certifications."""

    model_config = {"extra": "ignore"}

    name: Optional[str] = Field(None, max_length=100)


class LanguageItem(BaseModel):
    """One language as stored in canonical legacy profile.languages."""

    model_config = {"extra": "ignore"}

    name: Optional[str] = Field(None, max_length=100)
    proficiency: Optional[str] = Field(None, max_length=50)


class CareerProfile(BaseModel):
    """Read-only confirmed Career Profile derived from canonical legacy fields."""

    model_config = {"extra": "ignore"}

    skills: List[SkillItem] = Field(default_factory=list)
    certifications: List[CertificationItem] = Field(default_factory=list)
    languages: List[LanguageItem] = Field(default_factory=list)

File: apps/web/lib/schemas/careerProfile.ts
import { z } from "zod";

export const ExperienceItemSchema = z.object({
  id: z.string().nullable().optional(),
  role: z.string().max(200).nullable().optional(),
  company: z.string().max(200).nullable().optional(),
  start_date: z.string().max(20).nullable().optional(),
  end_date: z.string().max(20).nullable().optional(),
  description: z.string().max(2000).nullable().optional(),
  location: z.string().max(200).nullable().optional(),
  provenance: z.literal("extracted_from_cv").optional(),
  source_document_id: z.string().nullable().optional(),
  confidence: z.number().nullable().optional(),
  confirmed_at: z.string().nullable().optional(),
  updated_at: z.string().nullable().optional(),
}).passthrough();

export const EducationItemSchema = z.object({
  id: z.string().nullable().optional(),
  institution: z.string().max(200).nullable().optional(),
  degree: z.string().max(100).nullable().optional(),
  field: z.string().max(100).nullable().optional(),
  start_date: z.string().max(20).nullable().optional(),
  end_date: z.string().max(20).nullable().optional(),
  provenance: z.literal("extracted_from_cv").optional(),
  source_document_id: z.string().nullable().optional(),
  confidence: z.number().nullable().optional(),
  confirmed_at: z.string().nullable().optional(),
  updated_at: z.string().nullable().optional(),
}).passthrough();

export const SkillItemSchema = z.object({
  name: z.string().max(100).nullable().optional(),
}).strict();

export const CertificationItemSchema = z.object({
  name: z.string().max(100).nullable().optional(),
}).strict();

export const LanguageItemSchema = z.object({
  name: z.string().max(100).nullable().optional(),
  proficiency: z.string().max(50).nullable().optional(),
}).strict();

export const CareerProfileSchema = z.object({
  skills: z.array(SkillItemSchema).optional().default([]),
  certifications: z.array(CertificationItemSchema).optional().default([]),
  languages: z.array(LanguageItemSchema).optional().default([]),
}).strict();

export type ExperienceItem = z.infer<typeof ExperienceItemSchema>;
export type EducationItem = z.infer<typeof EducationItemSchema>;
export type SkillItem = z.infer<typeof SkillItemSchema>;
export type CertificationItem = z.infer<typeof CertificationItemSchema>;
export type LanguageItem = z.infer<typeof LanguageItemSchema>;
export type CareerProfile = z.infer<typeof CareerProfileSchema>;
